utils/src/main/java/software/amazon/awssdk/utils/AttributeMap.java

```diff
@@ -69,9 +88,15 @@ public <T> T get(Key<T> key) {
      * @return New options with values merged.
      */
     public AttributeMap merge(AttributeMap lowerPrecedence) {
-        Map<Key<?>, Object> copiedConfiguration = new HashMap<>(attributes);
-        lowerPrecedence.attributes.forEach(copiedConfiguration::putIfAbsent);
-        return new AttributeMap(copiedConfiguration);
+        Builder resultBuilder = new AttributeMap.Builder(this);
+        lowerPrecedence.attributes.forEach((k, v) -> {
+            resultBuilder.internalPutIfAbsent(k, () -> {
+                Value<?> result = v.copy();
+                result.clearCache();
```

Comment:
When we do `attributeMapBuilder.build().toBuilder().someModification(...)`, we `copy()` every value so that the modifications to the value don't affect the `attributes` of the past builder or attributeMap, but we want to preserve the cached values, because they're still valid.